### PR TITLE
feat(kuma-cp): support MeshHTTPRoute targetRef

### PR DESCRIPTION
**Pull request by lobkovilya**
Depends on https://github.com/kumahq/kuma/pull/6973

Implements [MADR](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/028-meshhttproute-as-targetref.md).

Implementation adds an artificial tag `__rule-matches-hash__` for subsets with HTTP routes. At the same time MeshHTTPRoute plugin generates routes with the name equal to `__rule-matches-hash__`. Thanks to that configurers are able to match place in envoy config with policy (see MeshTimeout configurer implementation).

### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues -- https://github.com/kumahq/kuma/issues/6923
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/core/matchers/testdata/matchedpolicies/dataplanepolicies/05.dataplane.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+type: Dataplane
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/core/matchers/testdata/matchedpolicies/dataplanepolicies/05.dataplane.yaml:1]: which test is this coming from? I think we should give test the files meaningful names

File: pkg/plugins/policies/core/matchers/dataplane.go [hunk 2]
```diff
@@ -70,37 +74,113 @@ func MatchedPolicies(rType core_model.ResourceType, dpp *core_mesh.DataplaneReso
...
+	}
+	return nil, errors.Errorf("MeshHTTPRoute %v not found", refName)
+}
+
+func isReferenced(refMeta core_model.ResourceMeta, refName string, resourceMeta core_model.ResourceMeta) bool {
+	if len(refMeta.GetNameExtensions()) == 0 {
+		return refName == resourceMeta.GetName()
+	}
+
+	if ns := refMeta.GetNameExtensions()[controllers.KubeNamespaceTag]; ns != "" {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/policies/core/matchers/dataplane.go:165]: It takes the namespace of the policy. If we have:
```yaml
kind: MeshTimeout
metadata:
  name: mt-1
  namespace: kuma-system
spec:
  targetRef:
    type: MeshHTTPRoute
    name: route-1
  to: []
```
then we'll reference `route-1` from `kuma-system`. 

Probably we'll have to add `namespace` to targetRef in the future when we'll support not only `kuma-system` namespaces in policies. It should be well thought and requires MADR, especially part with Universal integration.

File: pkg/plugins/policies/core/matchers/dataplane.go [hunk 3]
```diff
@@ -70,37 +74,113 @@ func MatchedPolicies(rType core_model.ResourceType, dpp *core_mesh.DataplaneReso
...
+		if isReferenced(refMeta, refName, item.GetMeta()) {
+			return item.(*meshhttproute_api.MeshHTTPRouteResource), nil
+		}
+	}
+	return nil, errors.Errorf("MeshHTTPRoute %v not found", refName)
+}
+
+func isReferenced(refMeta core_model.ResourceMeta, refName string, resourceMeta core_model.ResourceMeta) bool {
+	if len(refMeta.GetNameExtensions()) == 0 {
+		return refName == resourceMeta.GetName()
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/core/matchers/dataplane.go:162]: I'd rather expect to compare only name if `resourceMeta` doesn't have a `Namespace`, is this correct if `resourceMeta` has a `Namespace` but `refMeta` doesn't?

**Comment by michaelbeaumont** [comment]
Could this be organized into 2 PRs where we introduce support for targeting MesHHTTPRoute and one for adding it to MeshTimeout?

File: pkg/plugins/policies/core/matchers/dataplane.go [hunk 4]
```diff
@@ -70,37 +74,113 @@ func MatchedPolicies(rType core_model.ResourceType, dpp *core_mesh.DataplaneReso
...
+	}
+
+	if ns := refMeta.GetNameExtensions()[controllers.KubeNamespaceTag]; ns != "" {
+		return util_k8s.K8sNamespacedNameToCoreName(refName, ns) == resourceMeta.GetName()
+	}
+
+	return false
+}
+
+func unresolve(rl []core_model.Resource) []core_model.Resource {
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/core/matchers/dataplane.go:172]: is this function necessary? Isn't a ResolvedResource transparently a resource?

Comment (lobkovilya) [reply pkg/plugins/policies/core/matchers/dataplane.go:172]: Unfortunately no, there are plenty of places we're casting `core_model.Resource`:
image

File: pkg/plugins/policies/core/rules/rules.go [hunk 5]
```diff
@@ -246,6 +259,71 @@ func BuildToRules(matchedPolicies []core_model.Resource) (ToRules, error) {
...
+		mhr, ok = resolved.(*v1alpha1.MeshHTTPRouteResource)
+		if !ok {
+			return nil, errors.New("resolved MeshHTTPRoute policy has wrong type")
+		}
+	default:
+		return policyWithTo.GetToList(), nil
+	}
+
+	rv := []core_model.PolicyItem{}
+	for _, mhrRules := range mhr.Spec.To {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/policies/core/rules/rules.go:290]: I'm not sure, we're using `.Rules` in the next line, so it's quite specific to MeshHTTPRoute:
```go
for _, mhrRules := range mhr.Spec.To {
    for _, mhrRule := range mhrRules.Rules {
```

**Review by michaelbeaumont — APPROVED** [review]
LGTM, I guess little of this can be used for `kind: MeshGateway`? Since MeshHTTPRoute targets some Envoy resources whereas MeshGateway targets Dataplanes... :thinking:

File: pkg/plugins/policies/core/matchers/testdata/matchedpolicies/dataplanepolicies/05.dataplane.yaml (hunk 1 above)
Comment (lobkovilya) [reply pkg/plugins/policies/core/matchers/testdata/matchedpolicies/dataplanepolicies/05.dataplane.yaml]: moving this test to MeshTimeout PR

**Pull request merged**